### PR TITLE
[pigeon] Fix Kotlin generated sync host api error

### PR DESCRIPTION
**Pull request by 5peak2me**
the kotlin code only reply when it catch the error.

image

this is the java generation code

image

fixes https://github.com/flutter/flutter/issues/113187

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/pigeon/test/kotlin_generator_test.dart [hunk 1]
```diff
@@ -820,6 +942,24 @@ void main() {
...
+              val yArg = args[1].let { if (it is Int) it.toLong() else it as Long }
+              wrapped["result"] = api.add(xArg, yArg)
+            } catch (exception: Error) {
+              wrapped["error"] = wrapError(exception)
+            }
+            reply.reply(wrapped)
+          }
+        } else {
+          channel.setMessageHandler(null)
+        }
```
Comment (stuartmorgan) [inline packages/pigeon/test/kotlin_generator_test.dart:961]: There's a lot of redundant testing here; we should have specific, targeted tests for specific generation pieces, and only assert the relevant things in those tests. Testing a lot of unrelated things repeatedly makes test maintenance cumbersome and brittle.

File: packages/pigeon/lib/kotlin_generator.dart [hunk 2]
```diff
@@ -98,7 +98,7 @@ void _writeCodec(Indent indent, Api api, Root root) {
 
     indent.write(
         'override fun writeValue(stream: ByteArrayOutputStream, value: Any?) ');
-    indent.writeScoped('{', '}', () {
+    indent.scoped('{', '}', () {
```
Comment (stuartmorgan) [inline packages/pigeon/lib/kotlin_generator.dart:101]: How is this change related to the PR?

File: packages/pigeon/lib/kotlin_generator.dart [hunk 3]
```diff
@@ -628,7 +629,7 @@ void generateKotlin(KotlinOptions options, Root root, StringSink sink) {
         if (klass.fields.last != element) {
           indent.addln(',');
         } else {
-          indent.addln('');
+          indent.add('');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/kotlin_generator.dart:632]: Same question here.

File: packages/pigeon/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+## 4.2.3
+
+* [kotlin] Fixes Kotlin generation error when using the generate sync host api
```
Comment (stuartmorgan) [inline packages/pigeon/CHANGELOG.md:3]: This is missing the period required by the style guide.

I'm also having trouble parsing the last part of this sentence. I'm not sure what "generation error when using the generate [generated?] sync host api" means; generation doesn't happen when using the API. This also doesn't actually explain what the error is, just that there is one.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: packages/pigeon/lib/kotlin_generator.dart (hunk 2 above)
Comment (5peak2me) [reply packages/pigeon/lib/kotlin_generator.dart]: this change aim to format the code style like 
`override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {`
but the generated code is a char: `\t`. i have drop commit and open another pr later.

File: packages/pigeon/lib/kotlin_generator.dart (hunk 3 above)
Comment (5peak2me) [reply packages/pigeon/lib/kotlin_generator.dart]: this change reduce the blank line in data class.
```kotlin
data class Book (
  val title: String? = null,
  val author: String? = null

) {
```

File: packages/pigeon/test/kotlin_generator_test.dart [hunk 5]
```diff
@@ -126,6 +126,14 @@ void main() {
...
     expect(code, contains('fun doSomething(input: Input): Output'));
     expect(code, contains('channel.setMessageHandler'));
+    expect(code, contains('''
+            }
+            reply.reply(wrapped)
+          }
+        } else {
+          channel.setMessageHandler(null)
+        }
+    '''));
```
Comment (tarrinneal) [inline packages/pigeon/test/kotlin_generator_test.dart:136]: While this test does check to make sure this specific issue doesn't occur again, I'm not entirely sure this is useful. I'll defer to @stuartmorgan or @gaaclarke as to whether this should be removed, altered, or supplemented though.

**Review by tarrinneal — COMMENTED** [review]
This is a good fix, thank you for catching my error. I think this might point to an oversight in our e2e testing as well. I'll look into making that more robust.

File: packages/pigeon/test/kotlin_generator_test.dart (hunk 5 above)
Comment (5peak2me) [reply packages/pigeon/test/kotlin_generator_test.dart:136]: tks

**Comment by stuartmorgan** [comment]
What's the status of this PR? Is it ready for another review?

**Comment by 5peak2me** [comment]
> What's the status of this PR? Is it ready for another review?

yeah, Please.

**Review by tarrinneal — APPROVED** [review]

File: packages/pigeon/CHANGELOG.md (hunk 4 above)
Comment (5peak2me) [reply packages/pigeon/CHANGELOG.md]: done
